pkg/virt-handler/rest/console.go

```diff
@@ -0,0 +1,196 @@
+	}
+	t.stream(vmi, request, response, unixSocketPath, stopChn, cleanup)
+}
+
+func (t *ConsoleHandler) SerialHandler(request *restful.Request, response *restful.Response) {
+	vmi, code, err := t.getVMI(request)
+	if err != nil {
+		log.Log.Object(vmi).Reason(err).Error("Failed retrieve VMI")
+		response.WriteError(code, err)
+		return
+	}
+	unixSocketPath, err := t.getUnixSocketPath(vmi, "virt-serial0")
+	if err != nil {
+		log.Log.Object(vmi).Reason(err).Error("Failed unix socket for serial console")
+		response.WriteError(http.StatusBadRequest, err)
+		return
+	}
+	uid := vmi.GetUID()
+	stopCh := newStopChan(uid, t.serialLock, t.serialStopChans)
+	cleanup := func() {
+		deleteStopChan(uid, stopCh, t.serialLock, t.serialStopChans)
+	}
+	t.stream(vmi, request, response, unixSocketPath, stopCh, cleanup)
+}
+
+func (t *ConsoleHandler) getVMI(request *restful.Request) (*v1.VirtualMachineInstance, int, error) {
+	key := fmt.Sprintf("%s/%s", request.PathParameter("namespace"), request.PathParameter("name"))
+	vmiObj, vmiExists, err := t.vmiInformer.GetStore().GetByKey(key)
+	if err != nil {
+		return nil, http.StatusInternalServerError, err
+	}
+	if !vmiExists {
+		return nil, http.StatusNotFound, fmt.Errorf("VMI %s does not exist", key)
+	}
+	return vmiObj.(*v1.VirtualMachineInstance), 0, nil
+}
+
+func newStopChan(uid types.UID, lock *sync.Mutex, stopChans map[types.UID](chan struct{})) chan struct{} {
+	lock.Lock()
+	defer lock.Unlock()
+	// close current connection, if exists
+	if c, ok := stopChans[uid]; ok {
+		delete(stopChans, uid)
+		close(c)
+	}
+	// create a stop channel for the new connection
+	stopCh := make(chan struct{})
+	stopChans[uid] = stopCh
+	return stopCh
+}
+
+func deleteStopChan(uid types.UID, stopChn chan struct{}, lock *sync.Mutex, stopChans map[types.UID](chan struct{})) {
+	lock.Lock()
+	defer lock.Unlock()
+	// delete the stop channel from the cache if needed
+	if c, ok := stopChans[uid]; ok && c == stopChn {
+		delete(stopChans, uid)
+	}
+}
+
+func (t *ConsoleHandler) getUnixSocketPath(vmi *v1.VirtualMachineInstance, socketName string) (string, error) {
+	result, err := t.podIsolationDetector.Detect(vmi)
+	if err != nil {
+		return "", err
+	}
+	return path.Join("proc", strconv.Itoa(result.Pid()), "root", "var", "run", "kubevirt-private", string(vmi.GetUID()), socketName), nil
+}
+
+type cleanupOnError func()
+
+func (t *ConsoleHandler) stream(vmi *v1.VirtualMachineInstance, request *restful.Request, response *restful.Response, unixSocketPath string, stopCh chan struct{}, cleanup cleanupOnError) {
+	var upgrader = kubecli.NewUpgrader()
+	clientSocket, err := upgrader.Upgrade(response.ResponseWriter, request.Request, nil)
+	if err != nil {
+		log.Log.Object(vmi).Reason(err).Error("Failed to upgrade client websocket connection")
+		response.WriteError(http.StatusInternalServerError, err)
+		return
+	}
+	defer clientSocket.Close()
+
+	log.Log.Object(vmi).Infof("Websocket connection upgraded")
```

Comment:
If you want to keep these log lines, I think they should be `V(4)` (debug). In general we handle all errors, so if there is an error the errror log should reveal how fare  we go.

Reply:
Nah, those were remaining from the time before I wrote proper error-handling - removed